solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -895,3 +934,89 @@ def _name_pfxobs(current_names, forecast, i=1):
     else:
         return forecast_name
+
+
+def get_outage_periods(forecast, start, end, outages):
+    """Converts report outage periods to forecast data periods to
+    drop from analysis.
+
+    Parameters
+    ----------
+    forecast: Forecast
+    start: pd.Timestamp
+    end: pd.Timestamp
+    outages list of ReportOutages
+
+    Returns
+    -------
+    list of dicts
+        List of dicts with start and end keys. Times between these values
+        should not be included in analysis.
+    """
+    # First, determine a list of forecast issue times that include data that
+    # falls within the report
+
+    # Get total forecast horizon to know how far back to look for issue times
+    total_forecast_horizon = forecast.lead_time_to_start + forecast.run_length
+
+    # Look back another day, so we can start from a day's first issue time
+    # without undershooting the forecast horizon
+    lookback_period = total_forecast_horizon + pd.Timedelta('1D')
+
+    # Convert start to UTC so we can replace with the UTC-defined issue time
+    # to get the first issue time inside the lookback period.
+    lookback_start = start.tz_convert('UTC') - lookback_period
+    issue_start = lookback_start.replace(
+        hour=forecast.issue_time_of_day.hour
+    )
+
+    # Get all possible issue times that contribute to the report
+    issue_times = pd.date_range(
+        issue_start,
+        end.tz_convert('UTC'),
+        freq=forecast.run_length
+    )
+
+    outage_periods = []
+    # For each outage, if a forecast submission/issue_time falls within
+    # the outage. Create start/end bounds for the forecast data to exclude.
+    for outage in outages:
+        outage_submissions = issue_times[
+            (issue_times >= outage.start) & (issue_times <= outage.end)
+        ]
+        for issue_time in outage_submissions:
+            fx_start = issue_time + forecast.lead_time_to_start
+            fx_end = fx_start + forecast.run_length
+            outage_periods.append({
+                "start": fx_start,
+                "end": fx_end
+            })
+    return outage_periods
+
+
+def remove_outage_periods(outage_periods, data):
+    """Returns a copy of a dataframe with all values within an outage
+    period dropped.
+
+    Parameters
+    ----------
+    outage_periods: list of dict
+        List of dictionaries with start and end keys. Values should be
+        timestamps denoting the start and end of periods to remove.
+    data: pandas.DataFrame
+        The dataframe to drop outage data from.
+
+    Returns
+    -------
+    pd.DataFrame, int
+    """
+    if len(outage_periods) == 0:
+        return data, 0
+    dropped_total = 0
+    new_data = data.copy()
+    for outage in outage_periods:
+        outage_index = (new_data.index >= outage['start']) & (
+            new_data.index <= outage['end'])
+        new_data = new_data[~outage_index]
+        dropped_total += outage_index.sum()
```

Comment:
I don't see a problem with this nested view of `new_data`, but it's different than what I'm used to. `new_data` can be a view of a view of a view of a view etc... Might be better (or worse!) to create a single index and then slice, or to create chunks of the data and then append. What do you think?

Reply:
Maybe the `Index.union` method is our friend here.

Reply:
I've gone with just keeping track of the boolean series in  by `&`ing the `outage_index` on each iteration, and then indexing the data at the end of the for loop.